### PR TITLE
Add Service type "ExternalName" which results in CNAME DNS

### PR DESCRIPTION
**Pull request by therc**
ExternalName allows kubedns to return CNAME records for external
services. No proxying is involved.

First step for https://github.com/kubernetes/features/issues/33

See original issue at
https://github.com/kubernetes/kubernetes/issues/13748

No release note yet, that will come with the kubedns change.

``` release-note
NONE
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30599)

### CODE REVIEW

**Comment by therc** [comment]
I also changed the comments a bit to make the two types.go files use the same wording for the same fields.

**Comment by therc** [comment]
cc @kubernetes/kube-api

**Comment by therc** [comment]
I assume that it'd be a good idea to also have defaulting and validation in. :-) Feel free to reassign to @smarterclayton, @thockin or both, since they are the feature shepherds.

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -100,6 +100,14 @@ func ValidateDNS1123Label(value string, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
+func ValidateDNS1123Subdomain(value string, fldPath *field.Path) field.ErrorList {
```
Comment (smarterclayton) [inline pkg/api/validation/validation.go]: Godoc on this.

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -2139,9 +2148,15 @@ func ValidateService(service *api.Service) field.ErrorList {
 				allErrs = append(allErrs, field.Invalid(portPath, port.Port, "may not expose port 10250 externally since it is used by kubelet"))
 			}
 		}
+	} else if service.Spec.Type == api.ServiceTypeExternalName {
```
Comment (smarterclayton) [inline pkg/api/validation/validation.go]: would be better as a switch (else if usually means "use switch instead"

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -2139,9 +2148,15 @@ func ValidateService(service *api.Service) field.ErrorList {
...
+		if !isHeadlessService {
+			allErrs = append(allErrs, field.Invalid(specPath.Child("clusterIP"), service.Spec.ClusterIP, "must be 'None'"))
+		}
+		if len(service.Spec.ExternalName) > 0 {
```
Comment (smarterclayton) [inline pkg/api/validation/validation.go]: Is it allowed to have type ExternalName but an empty ExternalName?

**Comment by smarterclayton** [comment]
Given that this is alpha, we generally implement this as an annotation.  In this case the type is the new thing, so we should probably call it "AlphaExternalName".  The only concern there would be that because this is an enum that we'd potentially have services stored in the db without an enum, which could cause breaks.

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (therc) [inline pkg/api/validation/validation.go]: Done

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (therc) [inline pkg/api/validation/validation.go]: Done.

File: pkg/api/validation/validation.go (hunk 3 above)
Comment (therc) [inline pkg/api/validation/validation.go]: That was a question I had. I added  an error for empty names.

**Comment by thockin** [comment]
Review status: 0 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/api/types.go, line 1742 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPL9tj3UD_aIhO6AqDO:-KPL9tj3UD_aIhO6AqDP:brm8w7v) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2fa470740e300adfba042b46af2a6d9cc35fa90/pkg/api/types.go#L1742)):*
> ```Go
> // ServiceSpec describes the attributes that a user creates on a service
> type ServiceSpec struct {
> 	// Type determines how to expose the service. Defaults to ClusterIP and only
> ```

This makes it sound like "type" does not apply if clusterIP == "none", but type=ExternalName doesn't need clusterIP ?

---

*[pkg/api/v1/defaults.go, line 117 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPLBdYXfBJlOYS38njO:-KPLBdYYdxxZI73ovSac:b-4cmdxy) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2fa470740e300adfba042b46af2a6d9cc35fa90/pkg/api/v1/defaults.go#L117)):*
> ```Go
> 		obj.Type = ServiceTypeClusterIP
> 	} else if obj.Type == ServiceTypeExternalName {
> 		obj.ClusterIP = ClusterIPNone
> ```

I don't think you should default this here.  It should instead be a validation error if type is externalName but clusterIP != ""

Uggh, but that is in the REST path, isn't it?  By the time you get here, an IP has been allocated.  If you override this, you will leak.  Or did I get the sequence wrong?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599)*

**Comment by thockin** [comment]
yeah, I am nervous about how to extend an enum like this.  Can we do i
t as an annotation and say that they annotation is only considered if type=ClusterIP and clusterIP==None ?

---

Review status: 0 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-:-KPLDB5ItF9jg8TXtPMU:b6vge84)*

**Comment by smarterclayton** [comment]
We should maybe discuss the annotation thing on the api machinery call today.

**Comment by thockin** [comment]
I don't  know if there's a general answer to extending arbitrary enums with alpha values, especially when some platforms don't want to enable alpha features.  I think it has to be case-by-case.  In this case, define the requirement as the most innocuous setting + an alpha override.

Also, alpha features have to be gateable.  Do we need a --enable-alpha-annotations flag for apiserver that everyone can check?

---

Reviewed 1 of 5 files at r2, 14 of 14 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-:-KPODGKPpnN9zTDv13NA:b6hsbzf)*

**Comment by therc** [comment]
I might be able to join in the second half of the API meeting, but I'm at IDF16, so I'm not sure if I'll be able to find a corner quiet enough.

**Comment by smarterclayton** [comment]
We don't have anything gating alpha features today consistently.  Partially
that's because external components don't have shared global config.  I'd
probably argue that any alpha feature represented in the API needs to be
controlled by the API server, which potentially means filtering annotations
on both input and output.  The security issues with annotations are also
problematic - since annotations aren't discoverable (you can't discover
which alpha features exist) lots of implementers don't realize they need to
support / defend against annotations, and since annotations are one big
bucket there's no way to limit which annotations can get set by which users
short of admission control.

**Comment by thockin** [comment]
Something we DIDN'T do for this is open a feature bug.  @therc should do that.  Given your unavailability this week, and the lack of a corresponding DNS change, I have very little confidence it will make the 1.4 window. :(

**Comment by thockin** [comment]
It seems I am wrong, and we DO have a feature bug, but maybe have not been updating it.

**Comment by therc** [comment]
I mentioned the feature at the top. I can't update it, though.Oh, Github. :-)

I'm trying to finish the DNS change today.

**Comment by thockin** [comment]
Running down the feature checklist:

- disabled by default: not done.  We have been designing this as a "real" change not as an alpha.  We have the option of bypassing alpha for this, but even beta requires to be an annotation rather than a real field.  We have the option of bypassing beta too, if we are REALLY confident in the API for it.

- minimal testing.  This has no e2e test coverage.  It should be a relatively simple test, but it is absent.

- minimal docs.  This has no associated docs.

This scope of change falls into the grey area between "obviously not a feature" and "obviously a feature".  I feel bad that I didn't think of it as a feature early on, so we could have guided development via annotations.

---

Review status: 14 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-:-KPP6jDhGGv-P5JrNiEi:bh6fmyz)*

**Comment by therc** [comment]
Does "disabled by default" mean "opt-in" (admin must get involved to make changes in the control plane) or "dormant, API defaults must be overridden by user"? I assumed the latter.

**Comment by thockin** [comment]
disabled-by-default means that users can not use the feature unless
cluster-admin has designated the feature to be enabled.

**Comment by therc** [comment]
Review status: 14 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/api/types.go, line 1742 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPL9tj3UD_aIhO6AqDO:-KPROyR50H28nLo6TcMr:b-nuqbt2) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2fa470740e300adfba042b46af2a6d9cc35fa90/pkg/api/types.go#L1742)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> This makes it sound like "type" does not apply if clusterIP == "none", but type=ExternalName doesn't need clusterIP ?
</details>
I reworded it, but it's still a bit clumsy.

---

*[pkg/api/v1/defaults.go, line 117 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPLBdYXfBJlOYS38njO:-KPR65BjZGC4M23vUQhB:b-ci7tgh) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2fa470740e300adfba042b46af2a6d9cc35fa90/pkg/api/v1/defaults.go#L117)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> I don't think you should default this here.  It should instead be a validation error if type is externalName but clusterIP != ""
> 
> Uggh, but that is in the REST path, isn't it?  By the time you get here, an IP has been allocated.  If you override this, you will leak.  Or did I get the sequence wrong?
</details>
I thought that defaults happened before validation, but I think they happen before CREATION instead. pkg/api/rest/create.go:BeforeCreate calls Validate. And pkg/registry/service/rest:Create calls BeforeCreate, then allocates a VIP. The VIP is allocated only if clusterIP == "". I changed IsServiceIPRequested to short-circuit to False for ExternalName services so thatthey don't trigger an allocation, either, then I changed the check to make sure that clusterIP == "". 

I have also removed the default.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599)*

**Comment by therc** [comment]
I'm not sure what the consensus was on the API conference call today regarding the exact annotations to use, so it's still using the extra field/enum while getting it to a more polished state.

**Comment by thockin** [comment]
So the open issues are:

1) Was this all in vain because it needs to be an annotation
2) Where's the corresponding kube-dns change
3) e2e
4) docs

---

Reviewed 1 of 1 files at r4, 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/api/types.go, line 1742 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPL9tj3UD_aIhO6AqDO:-KPTmH7lul9m1eqVh8ea:b9612hb) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2fa470740e300adfba042b46af2a6d9cc35fa90/pkg/api/types.go#L1742)):*
<details><summary><i>Previously, therc (Rudi C) wrote…</i></summary>
> I reworded it, but it's still a bit clumsy.
</details>
How about:

// Type determines how the Service is exposed.  Defaults to ClusterIP.  Valid options are ExternalName, ClusterIP, NodePort, and LoadBalancer.
// "ExternalName" maps to the specified `externalName`.
// "ClusterIP" allocates a cluster-internal IP address for load-balancing to endpoints.  Endpoints are determined by the `selector` or if that is not specified, by manual construction of an Endpoints object.  If `clusterIP` is "None", no virtual IP is allocated and the endpoints are published as a set of endpoints rather than a stable IP.
// "NodePort" builds on ClusterIP and allocates a port on every node which routes to the `clusterIP`.
// "LoadBalancer" builds on NodePort and creates an external load-balancer (if supported in the current cloud) which routes to the `clusterIP`

---

*[pkg/api/v1/defaults.go, line 117 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPLBdYXfBJlOYS38njO:-KPUX3NbnbiWR0pqOMNb:bnsbu1g) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2fa470740e300adfba042b46af2a6d9cc35fa90/pkg/api/v1/defaults.go#L117)):*
<details><summary><i>Previously, therc (Rudi C) wrote…</i></summary>
> I thought that defaults happened before validation, but I think they happen before CREATION instead. pkg/api/rest/create.go:BeforeCreate calls Validate. And pkg/registry/service/rest:Create calls BeforeCreate, then allocates a VIP. The VIP is allocated only if clusterIP == "". I changed IsServiceIPRequested to short-circuit to False for ExternalName services so thatthey don't trigger an allocation, either, then I changed the check to make sure that clusterIP == "". 
> 
> I have also removed the default.
</details>
Ahh yes, gooooood

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-:-KPUXG4iHPSYdLFu6CMX:bemz4f0)*

**Comment by therc** [comment]
I was up until almost 1 last night, I must have forgotten to press Submit. :-( #30931 implements the kube-dns change, along with simple unit tests and e2e.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-:-KPUcMc7hOsBS85MsU3-:b9yn5nm)*

**Comment by therc** [comment]
Review status: 1 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/api/types.go, line 1742 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPL9tj3UD_aIhO6AqDO:-KPYaXZ5Fi54Or3txOR3:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2fa470740e300adfba042b46af2a6d9cc35fa90/pkg/api/types.go#L1742)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> How about:
> 
> // Type determines how the Service is exposed.  Defaults to ClusterIP.  Valid options are ExternalName, ClusterIP, NodePort, and LoadBalancer.
> // "ExternalName" maps to the specified `externalName`.
> // "ClusterIP" allocates a cluster-internal IP address for load-balancing to endpoints.  Endpoints are determined by the `selector` or if that is not specified, by manual construction of an Endpoints object.  If `clusterIP` is "None", no virtual IP is allocated and the endpoints are published as a set of endpoints rather than a stable IP.
> // "NodePort" builds on ClusterIP and allocates a port on every node which routes to the `clusterIP`.
> // "LoadBalancer" builds on NodePort and creates an external load-balancer (if supported in the current cloud) which routes to the `clusterIP`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599)*

**Comment by thockin** [comment]
Code LGTM.

I discussed this with a few folks and I think we can move this through without the usual beta cycle.  Rationale: it's already opt-in (no change unless users explicitly activate it), the implementation is simple, test coverage should be sufficient, and the crater is relatively small.

I picked a couple nits, so I can't quite LGTM this - fix those and ping me.  The subsequent DNS change also LGTM.  Needs docs and final steps from the feature checklist.  I'll also re-title it for relnote

+lgtm 

---

Reviewed 5 of 15 files at r6.
Review status: 5 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/api/validation/validation.go, line 2232 \[r6\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPZ--g-2RaVjawI4lgC:-KPZ--g-2RaVjawI4lgD:bwnqimg) ([raw file](https://github.com/kubernetes/kubernetes/blob/5ea4ad4d51601c632a8c80e3096dafac0dfc9e0c/pkg/api/validation/validation.go#L2232)):*
> ```Go
> 			allErrs = append(allErrs, ValidateDNS1123Subdomain(service.Spec.ExternalName, specPath.Child("externalName"))...)
> 		} else {
> 			allErrs = append(allErrs, field.Invalid(specPath.Child("externalName"), service.Spec.ExternalName, "externalName must not be empty"))
> ```

error string should not repeat field name

This should be handled as field.Required() - see elsewhere

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-:-KPYyYj0YsPFx7SEgB4i:bbj3wu7)*

**Comment by therc** [comment]
Review status: 4 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/api/validation/validation.go, line 2232 \[r6\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-KPZ--g-2RaVjawI4lgC:-KPZ3fOSTOW0ZDbJUi3D:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5ea4ad4d51601c632a8c80e3096dafac0dfc9e0c/pkg/api/validation/validation.go#L2232)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> error string should not repeat field name
> 
> This should be handled as field.Required() - see elsewhere
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599)*

**Comment by thockin** [comment]
LGTM +lgtm 

---

Reviewed 1 of 1 files at r7.
Review status: 5 of 15 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30599#-:-KPZcHpBrfpn3BTGdL6l:bz6ahrc)*

**Comment by therc** [comment]
@thockin could you add the milestone so it doesn't end up at the very bottom of the submit queue? I also don't know if the fact that it's an API change makes it any higher than a P3.

**Pull request merged**